File: src/chain-observers/providers/post-provider.ts
import { comment, TAccountName } from "@hiveio/wax";
import { WorkerBeeIterable } from "../../types/iterator";
import { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import { IOperationTransactionPair, OperationClassifier } from "../classifiers/operation-classifier";
import { DataEvaluationContext } from "../factories/data-evaluation-context";
import { ProviderBase } from "./provider-base";

export type TPostProvided<TAccounts extends Array<TAccountName>> = {
  [K in TAccounts[number]]: WorkerBeeIterable<IOperationTransactionPair<comment>>;
};

export interface IPostProviderData<TAccounts extends Array<TAccountName>> {
  posts: Partial<TPostProvided<TAccounts>>;
};

export interface IPostProviderOptions {
  authors: string[];
}

export class PostProvider<TAccounts extends Array<TAccountName> = Array<TAccountName>> extends ProviderBase<IPostProviderOptions> {
  public readonly authors = new Set<TAccountName>();

  public pushOptions(options: IPostProviderOptions): void {
    for(const account of options.authors)
      this.authors.add(account);
  }

  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [OperationClassifier]
  }

  public async provide(data: DataEvaluationContext): Promise<IPostProviderData<TAccounts>> {
    const result = {
      posts: {}
    } as IPostProviderData<TAccounts>;

    const accounts = await data.get(OperationClassifier);
    if (accounts.operationsPerType.comment)
      for(const operation of accounts.operationsPerType.comment) {
        if (operation.operation.parent_author !== "")
          continue;

        for(const account of this.authors) {
          if(operation.operation.author !== account)
            continue;

          if (!result.posts[account])
            result.posts[account] = [];

          result.posts[account].push({
            operation: operation.operation,
            transaction: operation.transaction
          });
        }
      }

    for(const account in result.posts)
      result.posts[account] = new WorkerBeeIterable(result.posts[account]);

    return result;
  }
}

File: python/workerbee/chain_observers/providers/post_provider.py
"""PostProvider — provides post (top-level comment) operations grouped by author.

Mirrors the PostProvider from src/chain-observers/providers/blog-content-provider.ts.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
from ..classifiers.operation_classifier import OperationClassifier
from .provider_base import ProviderBase

if TYPE_CHECKING:
    from ..factories.data_evaluation_context import DataEvaluationContext
    from ..payloads import OpBodyTransactionPair, PostsPayload


class PostProvider(ProviderBase):
    def __init__(self) -> None:
        self.authors: dict[str, None] = {}

    def push_options(self, options: Any) -> None:
        for account in options["authors"]:
            self.authors[account] = None

    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        return [OperationClassifier]

    async def provide(self, data: DataEvaluationContext) -> PostsPayload:
        posts: dict[str, list[OpBodyTransactionPair]] = {}

        operations = await data.get(OperationClassifier)
        comment_ops = operations["operations_per_type"].get("comment_operation")
        if comment_ops:
            for operation in comment_ops:
                # Posts have empty parent_author
                if operation["operation"]["parent_author"] != "":
                    continue

                author = operation["operation"]["author"]
                if author not in self.authors:
                    continue

                if author not in posts:
                    posts[author] = []

                posts[author].append(
                    {
                        "operation": operation["operation"],
                        "transaction": operation["transaction"],
                    }
                )

        return {"posts": posts}
